internal/v4/list.go

```diff
@@ -0,0 +1,151 @@
+
+package v4 // import "gopkg.in/juju/charmstore.v5-unstable/internal/v4"
+
+import (
+	"net/http"
+	"sync/atomic"
+
+	"github.com/juju/utils/parallel"
+	"gopkg.in/errgo.v1"
+	"gopkg.in/juju/charmrepo.v2-unstable/csclient/params"
+
+	"gopkg.in/juju/charmstore.v5-unstable/internal/charmstore"
+	"gopkg.in/juju/charmstore.v5-unstable/internal/router"
+)
+
+// GET list[?filter=value…][&include=meta][&sort=field[+dir]]
+// https://github.com/juju/charmstore/blob/v4/docs/API.md#get-list
+func (h *ReqHandler) serveList(_ http.Header, req *http.Request) (interface{}, error) {
+	lp, err := parseListParams(req)
+	if err != nil {
+		return "", err
+	}
+	return h.doList(lp, req)
+}
+
+// doList performs a listing specified by ListParams. If lp
+// specifies that additional metadata needs to be added to the results,
+// then it is added. It also filters the result based on ACLs.
+func (h *ReqHandler) doList(lp charmstore.ListParams, req *http.Request) (interface{}, error) {
+	// perform query
+	results, err := h.Store.List(lp)
+	if err != nil {
+		return nil, errgo.Notef(err, "error listing charms and bundles")
+	}
+	filteredACLResults := make([]*router.ResolvedURL, 0)
+	for _, result := range results.Results {
+		if err = h.AuthorizeEntity(result, req); err == nil {
+			filteredACLResults = append(filteredACLResults, result)
+		}
+	}
+	results.Results = filteredACLResults
+	return h.addMetaData(results, lp.Include, req)
+}
+
+//addMetada adds the requested meta data with the include list.
+func (h *ReqHandler) addMetaData(results charmstore.ListResult, include []string, req *http.Request) (interface{}, error){
+	response := params.ListResponse{
+		Results:    make([]params.EntityResult, len(results.Results)),
+	}
+	run := parallel.NewRun(maxConcurrency)
+	var missing int32
+	for i, ref := range results.Results {
+		i, ref := i, ref
+		run.Do(func() error {
+			meta, err := h.Router.GetMetadata(ref, include, req)
+			if err != nil {
+				// Unfortunately it is possible to get errors here due to
+				// internal inconsistency, so rather than throwing away
+				// all the search results, we just log the error and move on.
+				logger.Errorf("cannot retrieve metadata for %v: %v", ref, err)
+				atomic.AddInt32(&missing, 1)
+				return nil
+			}
+			response.Results[i] = params.EntityResult{
+				Id:   ref.PreferredURL(),
+				Meta: meta,
+			}
+			return nil
+		})
+	}
+	// We never return an error from the Do function above, so no need to
+	// check the error here.
+	run.Wait()
+	if missing == 0 {
+		return response, nil
+	}
+	// We're missing some results - shuffle all the results down to
+	// fill the gaps.
+	j := 0
+	for _, result := range response.Results {
+		if result.Id != nil {
+			response.Results[j] = result
+			j++
+		}
+	}
+	response.Results = response.Results[0:j]
+	return response, nil
+}
+
+// parseListParms extracts the list parameters from the request
+func parseListParams(req *http.Request) (charmstore.ListParams, error) {
```

Comment:
This function is almost identical to parseSearchParams. ISTM that there's so much in common between the sort params and the list params that we should probably be using the same code to parse them both and only after parsing check for valid fields and map to the underlying entity names. The specific thing that tips me over the balance here is that I think it would be wrong if the two endpoints had a search field that they held in common but that had different semantics in each. That is, to the end user, the interface should be identical where possible. By having two separate parsers, we're inviting the implementations to diverge.

How about having a single parseSearchParams function (or other name that isn't either "search" or "list" :]) that can parse everything acceptable to both search and list, but accepts a set of allowed parameters. The field mapping that currently happens in ParseSortFields could be done in createSort, so the result of parseSearchParams would be independent of whether it would be used for search or list.